src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,251 @@
+                entitySet.id, entitySet.expiration, OffsetDateTime.now(), entitySet.expiration.deleteType, expirationPT
+        )
+        while (entityKeyIds.iterator().hasNext()) {
+            val chunkedEntityKeyIds = Iterables.partition(entityKeyIds, MAX_BATCH_SIZE)
+            for (idsChunk in chunkedEntityKeyIds) {
+                val ids = idsChunk.toSet()
+
+                val writeEvent = if (entitySet.expiration.deleteType == DeleteType.Hard) {
+                    if (!edm.isAssociationEntitySet(entitySet.id)) {
+                        deleteAssociationsOfExpiredEntities(entitySet.id, ids)
+                    }
+                    dataGraphService.deleteEntities(
+                            entitySet.id,
+                            ids,
+                            propertyTypes)
+
+                } else {
+                    if (!edm.isAssociationEntitySet(entitySet.id)) {
+                        clearAssociationsOfExpiredEntities(entitySet.id, ids)
+                    }
+                    dataGraphService.clearEntities(
+                            entitySet.id,
+                            ids,
+                            propertyTypes)
+                }
+
+                logger.info("Completed deleting {} expired elements from entity set {}.",
+                        writeEvent.numUpdates,
+                        entitySet.name)
+
+                totalDeletedEntitiesCount += writeEvent.numUpdates
+                deletedEntityKeyIds.addAll(ids)
+            }
+
+            auditingManager.recordEvents(
+                    listOf(AuditableEvent(
+                            IdConstants.SYSTEM_ID.id,
+                            AclKey(entitySet.id),
+                            AuditEventType.DELETE_EXPIRED_ENTITIES,
+                            "Expired entities deleted through BackgroundExpiredDataDeletionService",
+                            Optional.of(deletedEntityKeyIds),
+                            ImmutableMap.of(),
+                            OffsetDateTime.now(),
+                            Optional.empty())
+                    )
+            )
+        }
+        return totalDeletedEntitiesCount
+    }
+
+    private fun deleteAssociationsOfExpiredEntities(entitySetId: UUID, ids: Set<UUID>): List<WriteEvent> {
+        val associationEdgeKeys: PostgresIterable<DataEdgeKey> = dataGraphService.getEdgesConnectedToEntities(entitySetId, ids, true)
+        val associationPropertyTypes = associationEdgeKeys
+                .map {
+                    it.edge.entitySetId to
+                            edm.getPropertyTypesAsMap(
+                                    edm.getEntityType(edm.getEntitySet(it.edge.entitySetId).entityTypeId
+                                    ).properties
+                            )
+                }.toMap()
+        return dataGraphService.deleteAssociationsBatch(entitySetId, associationEdgeKeys, associationPropertyTypes)
```

Comment:
I feel like this will break because `PostgresIterable` is usually a one-use iterator (I think 🤔 )

Reply:
I took most of this logic from DataController.deleteAssociations, in which a PostgresIterable<DataEdgeKey> is streamed and then passed into another function. If that works this should work right?

Reply:
😨 
Then I'm guessing that `DataController.deleteAssociations` does not work correctly 🤔 
But does this code work for you locally? I would expect it to throw an exception at line 204

Reply:
It looks like PostgresIterable doesn't actually execute until you iterate, so you can iterate through it more than once with the query being executed each time you iterate... not sure this isn't leading to weird side effects in DataController

Reply:
So yeah, the iterating-only-once thing was a separate bug that I conflated with this due to their highly similar names. That being said, we have to make sure that when we get PostgresIterables as return values that we only iterate over them once because each iteration is a network roundtrip + postgres hit + open connection on postgres, all of which is muuuuch slower than doing something in memory.

Reply:
SO! I would recommend doing this instead:
```
val filteredAssociationEdgeKeys = associationEdgeESIds
    .filter { auditEdgeEntitySetIds.contains(it.) }
    .toSet()
```

Reply:
So I'd recommend using `associationEdgeESIds` instead of `associationEdgeKeys` on line 214. Also it looks like on 218 you're only using `it.edge.entitySetId` anyway so no need for `filteredAssociationEdgeKeys` to contain the entire `DataEdgeKey`

Reply:
i made it so the postgresiterable is collected into a set and deleteAssociationsBatch now just takes an iterable. But the usage in DataController is still dangerous.

Reply:
So I have some thoughts on this:
- If an API returns a `PostgresIterable`, I think that should be taken as an indication that the scale of data returned has a chance of being very large and therefore storing all in memory would be very bad
- It may not be the case with `getEdgesConnectedToEntities` but I think we should treat it that way for now

Reply:
So to wrap this up, I think it's best not to use a `Set<>` here for now just in case. I mean in theory you could have a full mesh graph subset of an entity-set and that call would return a toooooooooooon of results.
We'll go through and audit uses of `PostgresIterable` to make sure they're warranted

Reply:
okay I'll leave it as a pgiterable for now then